/*
 * FeatureA, exploit symbolic rewriting to see that a computation
 * failed.
 *
 * This is shown below. 'f' is a partially defined function which 
 * only rewrites '0'; i.e., 'f 2' rewrites to 'f 2' since 'f' isn't
 * defined on '2'. You can subsequently catch that a computation
 * didn't reduce to an integer, and throw that expression as an
 * exception.
 */

using System

def f = [ 0 -> 1 ]

def main = [ X: int -> X | X -> throw X ] (f 2)
